crates/bevy_reflect/src/type_data.rs

```diff
@@ -0,0 +1,166 @@
+///     "woof"
+///   }
+/// }
+/// ```
+///
+/// To do this, we can create a type data struct that implements [`TypeData`]:
+///
+/// ```
+/// # use bevy_reflect::{TypeData, Reflect, FromReflect};
+/// # trait Animal {
+/// #   fn speak(&self) -> &'static str;
+/// # }
+/// #
+/// # #[derive(Reflect)]
+/// # struct Dog(String);
+/// #
+/// # impl Animal for Dog {
+/// #   fn speak(&self) -> &'static str {
+/// #     "woof"
+/// #   }
+/// # }
+/// #
+/// #[derive(Clone)]
+/// struct ReflectAnimal {
+///   speak: fn(&dyn Reflect) -> &'static str,
+/// }
+///
+/// impl ReflectAnimal {
+///   pub fn speak(&self, animal: &dyn Reflect) -> &'static str {
+///     (self.speak)(animal)
+///   }
+/// }
+///
+/// impl<T: Animal + FromReflect> TypeData<T> for ReflectAnimal {
+///   fn create_type_data() -> Self {
+///     Self {
+///       speak: |animal| {
+///         T::from_reflect(animal).unwrap().speak()
+///       }
+///     }
+///   }
+/// }
+///
+/// // Usage
+/// let dog = Dog("Fido".to_string());
+/// let data = <ReflectAnimal as TypeData<Dog>>::create_type_data();
+/// assert_eq!(data.speak(&dog), "woof");
+/// ```
+///
+/// Alternatively, we can use the [`reflect_trait`] macro to generate a type data struct for us.
+/// Note that this only works with [object-safe] traits.
+///
+/// ```
+/// # use bevy_reflect::{TypeData, Reflect, FromReflect};
+/// # use bevy_reflect_derive::reflect_trait;
+/// #[reflect_trait]
+/// trait Animal {
+///   fn speak(&self) -> &'static str;
+/// }
+///
+/// # #[derive(Reflect)]
+/// # struct Dog(String);
+/// #
+/// # impl Animal for Dog {
+/// #   fn speak(&self) -> &'static str {
+/// #     "woof"
+/// #   }
+/// # }
+/// #
+/// // Usage
+/// let dog = Dog("Fido".to_string());
+/// let data = <ReflectAnimal as TypeData<Dog>>::create_type_data();
+/// assert_eq!(data.get(&dog).unwrap().speak(), "woof");
+/// ```
+///
+/// [derive macro]: bevy_reflect_derive::Reflect
+/// [`ReflectDefault`]: crate::std_traits::ReflectDefault
+/// [`reflect_trait`]: crate::reflect_trait
+/// [object-safe]: https://doc.rust-lang.org/reference/items/traits.html#object-safety
+pub trait TypeData<T>: BaseTypeData + Clone {
+    /// Create a new instance of this type data.
+    fn create_type_data() -> Self;
+
+    /// Callback for when the type data is fully registered.
+    ///
+    /// Type data becomes fully registered in one of the following ways:
+    /// * The containing [`TypeRegistration`] is inserted into a [`TypeRegistry`]
+    /// * This type data is inserted with [`TypeRegistry::register_type_data`]
+    ///
+    /// This can be used to register additional type data when this type data is registered.
+    /// For example, to register the same data for related types or to register it
+    /// for container types (e.g. `Vec<T>`, `Option<T>`, etc).
+    ///
+    /// [`TypeRegistration`]: crate::TypeRegistration
+    #[allow(unused_variables)]
+    fn on_register(registry: &mut TypeRegistry) {}
```

Comment:
What about renaming `on_register` to `register` and accept a `self` and the default impl just calls `registry.register_type_data`

Reply:
i think perhaps the signature should be `fn(&self, registry: &mut TypeRegistry)`, but i think taking `self` would make this API quite unpredictable. currently, its relationship with the methods on the type registry is clearly defined, and i wouldn't want to muddy the waters.

Reply:
I don't think the default impl should call `registry.register_type_data`, since that method will likely need to call this one.